idea/src/org/jetbrains/kotlin/idea/quickfix/sealedSubClassToObject/ConvertSealedSubClassToObjectFix.kt

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.quickfix.sealedSubClassToObject
+
+import com.intellij.codeInspection.LocalQuickFix
+import com.intellij.codeInspection.ProblemDescriptor
+import com.intellij.lang.Language
+import com.intellij.openapi.project.Project
+import com.intellij.psi.JavaPsiFacade
+import com.intellij.psi.PsiElement
+import com.intellij.psi.impl.source.tree.JavaElementType
+import com.intellij.psi.search.searches.ReferencesSearch
+import org.jetbrains.kotlin.KtNodeTypes
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.getParentOfType
+
+class ConvertSealedSubClassToObjectFix : LocalQuickFix {
+
+    override fun getFamilyName() = "Convert Sealed Sub-class to Object"
+
+    companion object {
+        val JAVA_LANG = Language.findLanguageByID("JAVA")
+        val KOTLIN_LANG = Language.findLanguageByID("kotlin")
+    }
+
+    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+        val klass = descriptor.psiElement.getParentOfType<KtClass>(false) ?: return
+
+        changeInstances(klass)
+        changeDeclaration(klass)
+    }
+
+    /**
+     * Changes declaration of class to object.
+     */
+    private fun changeDeclaration(element: KtClass) {
+        val factory = KtPsiFactory(element)
+
+        element.removeOpen()
+        element.changeToObject(factory)
+        element.transformToObject(factory)
+    }
+
+    private fun KtClass.removeOpen() {
+        modifierList?.getModifier(KtTokens.OPEN_KEYWORD)?.delete()
```

Comment:
Here please use `addModifier(FINAL_KEYWORD)` directly on `KtClass`. This function includes necessary logic to delete incompatible modifiers (like open, abstract or sealed) and not to add redundant modifiers (like final in most cases).